src/main/ApplicationImpl.cpp

```diff
@@ -552,27 +553,224 @@ ApplicationImpl::joinAllThreads()
+        }
+    }
+
+    return userSpecifiedSeqNum ? *userSpecifiedSeqNum : nextLedgerSeq;
+}
+
+void
+ApplicationImpl::setManualCloseVirtualTime(
+    optional<TimePoint> const& userSpecifiedCloseTime)
+{
+    TimePoint const maxSystemTimePointSeconds =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            VirtualClock::system_time_point::duration::max())
+            .count();
+    TimePoint const maxTimeT = std::numeric_limits<std::time_t>::max();
+    TimePoint const maxCloseTime =
+        std::min(maxSystemTimePointSeconds, maxTimeT) -
+        Herder::MAX_TIME_SLIP_SECONDS.count();
+
+    TimePoint const lastCloseTime = getLedgerManager()
+                                        .getLastClosedLedgerHeader()
+                                        .header.scpValue.closeTime;
+    if (lastCloseTime >= maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("Close time already at max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    auto systemNow = getClock().system_now();
+    TimePoint secondsSinceEpoch =
+        std::chrono::duration_cast<std::chrono::seconds>(
+            systemNow.time_since_epoch())
+            .count();
+    if (secondsSinceEpoch > maxCloseTime)
+    {
+        throw std::invalid_argument(
+            fmt::format(FMT_STRING("System time already past max ({}); can not "
+                                   "close further ledgers"),
+                        maxCloseTime));
+    }
+
+    uint64_t const now = VirtualClock::to_time_t(systemNow);
+    uint64_t nextCloseTime = std::max(now, lastCloseTime + 1);
+
+    if (userSpecifiedCloseTime)
+    {
+        if (*userSpecifiedCloseTime > maxCloseTime)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Specified close time past max ({})"),
+                            maxCloseTime));
+        }
+        if (*userSpecifiedCloseTime < nextCloseTime)
+        {
+            throw std::invalid_argument(
+                fmt::format(FMT_STRING("Invalid manual close time {} (last "
+                                       "close time={}, now={})"),
+                            *userSpecifiedCloseTime, lastCloseTime, now));
+        }
+        nextCloseTime = *userSpecifiedCloseTime;
+    }
+    else
+    {
+        uint64_t const defaultManualCloseTimeInterval =
+            getConfig().getExpectedLedgerCloseTime().count();
+        if (maxCloseTime - defaultManualCloseTimeInterval >= lastCloseTime)
+        {
+            uint64_t const defaultNextCloseTime =
+                lastCloseTime + defaultManualCloseTimeInterval;
+            nextCloseTime = std::max(nextCloseTime, defaultNextCloseTime);
+        }
+    }
+
+    auto const nextCloseSystemTime = VirtualClock::from_time_t(nextCloseTime);
+    if (VirtualClock::to_time_t(nextCloseSystemTime) != nextCloseTime)
```

Comment:
I know you have already removed these checks, but I want to emphasize that they don't make that much sense to begin with. As I was saying in my previous comment, `time_t` can be signed so you simply _cannot_ overflow it under any circumstances. You can't try to retroactively catch it, because overflowing it is already undefined behavior.

As an analogy, the following code is wrong
```
bool doesAdditionOverflow(int x, int y)
{
    int z = x + y;
    return (z - y) != x;
}
```
because the undefined behavior has already occurred by the time you maybe noticed (and I say maybe, because nothing is guaranteed!).

There is no action required on this, so feel free to resolve it once you have read it.

Reply:
Makes sense, thanks.  Resolving.